include/irr/asset/CTexturePacker.h

```diff
@@ -58,40 +58,46 @@ class CSquareAddressAllocator
 class ITexturePacker
 {
 protected:
-    _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_LAYER_SHIFT = 10u;//10bit 2d page address ( 32=2^(10/2) pages per dimension ) and up to 6bit for 3rd address dimension (layer)
-    _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_MORTON_MASK = (1u<<ADDR_LAYER_SHIFT)-1u;
+    const uint32_t ADDR_LAYER_SHIFT;
+    const uint32_t ADDR_MORTON_MASK;
+    const uint32_t TEX_OFFSET_X_MASK;
 
     const uint32_t m_pgSzxy;
     const uint32_t m_tilesPerDim;
     CSquareAddressAllocator<uint32_t> m_pgtAddrAlctr;
 
 public:
+    using page_tab_offset_t = core::vector2du32_SIMD;
+    static page_tab_offset_t page_tab_offset_invalid() { return page_tab_offset_t(~0u,~0u,~0u); }
+
+    //! STexOffset is what is stored in texels of page table!
     struct STexOffset
     {
-        _IRR_STATIC_INLINE_CONSTEXPR uint32_t X_MASK = ((1u<<(ADDR_LAYER_SHIFT>>1))-1u);
         _IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_addr = 0xffffu;
 
         STexOffset(uint32_t _addr) : addr(_addr) {}
-        STexOffset(uint32_t _x, uint32_t _y, uint32_t _layer)
-        {
-            addr = (_x&X_MASK);
-            addr |= (_y&X_MASK)<<(ADDR_LAYER_SHIFT>>1);
-            addr |= (_layer<<ADDR_LAYER_SHIFT);
-        }
-        uint32_t x() const { return addr&X_MASK; }
-        uint32_t y() const { return (addr>>(ADDR_LAYER_SHIFT>>1))&X_MASK; }
-        uint32_t layer() const { return (addr&0xffffu)>>ADDR_LAYER_SHIFT; }
-
-        bool isValid() const { return (addr&0xffffu)==invalid_addr; }
-        bool hasMipTailAddr() const { return addr&(0xffffu<<16); }
-        STexOffset getMipTailAddr() const { return addr>>16; }
 
         //upper 16 bits are used for address of mip-tail page
         uint32_t addr;
     };
+    STexOffset makeTexOffset(uint32_t _x, uint32_t _y, uint32_t _layer)
```

Comment:
dont call it `STexOffset` if this is your pointer to the physical storage page

a name TextureOffset is only allowed to name the offset into virtual address space